layers/cmd/ingest_csv.go

```diff
@@ -30,6 +31,93 @@ import (
 	"github.com/cloudprivacylabs/opencypher/graph"
 )
 
+type CSVIngester struct {
+	BaseIngestParams
+	StartRow     int
+	EndRow       int
+	HeaderRow    int
+	ID           string
+	InitialGraph string
+}
+
+func (ci CSVIngester) Run(pipeline *PipelineContext) error {
+	// pipeline.Input = strings.NewReader(context.Background().Value("filename").(string))
+	initialGraph := ci.InitialGraph
+	layer, err := LoadSchemaFromFileOrRepo(pipeline.Context, ci.CompiledSchema, ci.Repo, ci.Schema, ci.Type, ci.Bundle)
+	if err != nil {
+		return err
+	}
+
+	reader := csv.NewReader(pipeline.Input)
+	startRow := ci.StartRow
+	endRow := ci.EndRow
+	headerRow := ci.HeaderRow
+	if headerRow >= startRow {
+		return errors.New("Header row is ahead of start row")
+
+	}
+	var grph graph.Graph
+	if layer != nil && initialGraph != "" {
+		grph = pipeline.Graph
+	} else {
+		grph = ls.NewDocumentGraph()
+	}
+	embedSchemaNodes := ci.EmbedSchemaNodes
+	onlySchemaAttributes := ci.OnlySchemaAttributes
+
+	parser := csvingest.Parser{
+		OnlySchemaAttributes: onlySchemaAttributes,
+		SchemaNode:           layer.GetSchemaRootNode(),
+	}
+
+	builder := ls.NewGraphBuilder(grph, ls.GraphBuilderOptions{
+		EmbedSchemaNodes:     embedSchemaNodes,
+		OnlySchemaAttributes: onlySchemaAttributes,
+	})
+
+	idTemplate := ci.ID
+	idTmp, err := template.New("id").Parse(idTemplate)
+	if err != nil {
+		return err
+	}
+	for row := 0; ; row++ {
+		rowData, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if headerRow == row {
+			parser.ColumnNames = rowData
+		} else if row >= startRow {
+			if endRow != -1 && row > endRow {
+				break
+			}
+			templateData := map[string]interface{}{
+				"rowIndex":  row,
+				"dataIndex": row - startRow,
+				"columns":   rowData,
+			}
+			buf := bytes.Buffer{}
+			if err := idTmp.Execute(&buf, templateData); err != nil {
+				return err
+			}
+
+			parsed, err := parser.ParseDoc(pipeline.Context, strings.TrimSpace(buf.String()), rowData)
+			if err != nil {
+				return err
+			}
+			_, err = ls.Ingest(builder, parsed)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
For each row ingested. call pipelinecontext.Next(). See comments for pipelineContext